### PR TITLE
Data Generation Upgrade- Speed, Performance and Coverage

### PR DESCRIPTION
**Pull request by lior-dikstein**
#### API Changes
- `output_image_size` and `extra_pixels` now accept `int` or `tuple`.

#### Image Pipeline
- Added new `SmoothAugmentationImagePipeline`, replacing old pipeline in PyTorch.
- Set `SmoothAugmentationImagePipeline` as default for PyTorch/TensorFlow.

#### Performance
- Added float16 support in PyTorch (~2x speedup).
- Updated data generation defaults for PyTorch/TensorFlow.

#### Layer and Scheduler Updates
- Aligned TensorFlow `ReduceLROnPlateau` with PyTorch.
- Added `ReduceLROnPlateauWithReset` scheduler in TensorFlow/PyTorch.

#### Miscellaneous
- Extracted grid creation into separate function.
- Fixed variable name misspellings.
- Renamed and added output loss functions.
- Updated constants and default values.
- Added function to clip float16 max values.
- Supported BN layers with `None` activations.

#### Bug Fixes
- Fixed `image_clipping` in PyTorch and Keras.

#### Tests
- Added tests for new pipelines, schedulers, and loss functions.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/data_generation/keras/optimization_functions/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+            raise ValueError(f'mode {mode} is unknown!')
+        if threshold_mode not in {'rel', 'abs'}:
+            raise ValueError(f'threshold mode {threshold_mode} is unknown!')
+
+        if mode == 'min':
+            self.mode_worse = float('inf')
+        else:  # mode == 'max':
+            self.mode_worse = float('-inf')
+
+        self.mode = mode
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/keras/optimization_functions/lr_scheduler.py:183]: Occures in init no?

File: model_compression_toolkit/data_generation/keras/optimization_functions/lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+        self.threshold = threshold
+        self.threshold_mode = threshold_mode
+
+    def get_config(self) -> Dict:
+        """
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/keras/optimization_functions/lr_scheduler.py:363]: Need also in ReduceLROnPlateau?

File: model_compression_toolkit/data_generation/pytorch/image_operations.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+        Args:
+            size (int): The size of the Gaussian kernel.
+            sigma (float): The standard deviation of the Gaussian kernel.
+            kernel (torch.Tensor, optional): Precomputed Gaussian kernel. If None, it will be created.
+        """
+        super().__init__()
+        if kernel is None:
+            kernel = self.gaussian_kernel(size, sigma)
+        kernel = kernel.view(1, 1, kernel.shape[0], kernel.shape[1])
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/image_operations.py:52]: check kernel shape?

File: model_compression_toolkit/data_generation/pytorch/image_operations.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    """
+    A PyTorch module for applying Gaussian smoothing to an image.
+    """
+
+    def __init__(self, size: int = 3, sigma: float = 1.25, kernel: torch.Tensor = None):
+        """
+        Initialize the Smoothing module.
+
+        Args:
+            size (int): The size of the Gaussian kernel.
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/image_operations.py:45]: Maybe state default values

File: model_compression_toolkit/data_generation/pytorch/image_operations.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+# ==============================================================================
+import numpy as np
+import torch
+import torch.nn.functional as F
+from model_compression_toolkit.core.pytorch.pytorch_device_config import get_working_device
+from torchvision.transforms import Normalize
+
+def create_valid_grid(means, stds) -> torch.Tensor:
+    """
+    Create a valid grid for image normalization.
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/image_operations.py:23]: Add type hints

File: model_compression_toolkit/data_generation/pytorch/image_operations.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        Initialize the Smoothing module.
+
+        Args:
+            size (int): The size of the Gaussian kernel.
+            sigma (float): The standard deviation of the Gaussian kernel.
+            kernel (torch.Tensor, optional): Precomputed Gaussian kernel. If None, it will be created.
+        """
+        super().__init__()
+        if kernel is None:
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/image_operations.py:50]: if kernel is provided you don't need size and sigma?

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/bn_layer_weighting_functions.py [hunk 7]
```diff
@@ -18,30 +18,40 @@
 
 from model_compression_toolkit.data_generation.common.enums import BNLayerWeightingType
 from model_compression_toolkit.data_generation.pytorch.model_info_exctractors import OriginalBNStatsHolder, \
-    ActivationExtractor
+    ActivationExtractor, PytorchActivationExtractor
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/optimization_functions/bn_layer_weighting_functions.py:21]: Are you using ActivationExtractor here?

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        if not isinstance(optimizer, Optimizer):
+            raise TypeError('{} is not an Optimizer'.format(
+                type(optimizer).__name__))
+        self.optimizer = optimizer
+
+        if isinstance(min_lr, (list, tuple)):
+            if len(min_lr) != len(optimizer.param_groups):
+                raise ValueError("expected {} min_lrs, got {}".format(
+                    len(optimizer.param_groups), len(min_lr)))
+            self.min_lrs = list(min_lr)
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py:67]: It won't make problems if min_lr is a list?

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+            raise ValueError('mode ' + mode + ' is unknown!')
+        if threshold_mode not in {'rel', 'abs'}:
+            raise ValueError('threshold mode ' + threshold_mode + ' is unknown!')
+
+        if mode == 'min':
+            self.mode_worse = float('inf')
+        else:  # mode == 'max':
+            self.mode_worse = float('-inf')
+
+        self.mode = mode
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py:204]: The values changed since init?

File: model_compression_toolkit/core/pytorch/constants.py [hunk 10]
```diff
@@ -94,3 +94,8 @@
 # # Batch size value for 'reshape' and 'view' operators,
 # # the value is -1 so the batch size is inferred from the length of the array and remaining dimensions.
 BATCH_DIM_VALUE = -1
+
+
+# The maximum and minimum representable values for float16
+MAX_FLOAT16 = torch.finfo(torch.float16).max - 1
+MIN_FLOAT16 = torch.finfo(torch.float16).min - 1
```
Comment (lapid92) [inline model_compression_toolkit/core/pytorch/constants.py:101]: Are you using the MIN_FLOAT16?

File: model_compression_toolkit/data_generation/common/constants.py [hunk 11]
```diff
@@ -17,6 +17,9 @@
 # Define a constant for the image input key.
 IMAGE_INPUT = 'image_input'
 
+# Output string
+OUTPUT = 'output'
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/common/constants.py:21]: Are you using the OUTPUT const?

**Review by lapid92 — APPROVED** [review]

File: model_compression_toolkit/data_generation/pytorch/image_operations.py (hunk 6 above)
Comment (lior-dikstein) [reply model_compression_toolkit/data_generation/pytorch/image_operations.py:50]: yes

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/bn_layer_weighting_functions.py (hunk 7 above)
Comment (lior-dikstein) [reply model_compression_toolkit/data_generation/pytorch/optimization_functions/bn_layer_weighting_functions.py:21]: In theory, you can. That is what we did in the research. Here we only have one function.

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py (hunk 8 above)
Comment (lior-dikstein) [reply model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py:67]: I copied this part from Pytorch so I think it is ok

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py (hunk 9 above)
Comment (lior-dikstein) [reply model_compression_toolkit/data_generation/pytorch/optimization_functions/lr_scheduler.py:204]: Fixed it

File: model_compression_toolkit/core/pytorch/constants.py (hunk 10 above)
Comment (lior-dikstein) [reply model_compression_toolkit/core/pytorch/constants.py:101]: Yes

File: tests/data_generation_tests/keras/test_scheduler_step.py [hunk 12]
```diff
@@ -17,60 +17,62 @@
...
-        self.opt_lr.learning_rate = 0.1
+        self.optimizer.learning_rate = 0.1
         self.scheduler._reset()
-        self.scheduler.on_epoch_end(0.1)  # No improvement
-        for _ in range(self.scheduler.patience):
-            self.scheduler.on_epoch_end(0.1)
-        self.assertLess(float(self.opt_lr.learning_rate.numpy()), 0.1)
+        for p in range(self.scheduler.patience + 2):
+            self.scheduler.on_epoch_end(p, 0.1)
+        self.assertLess(float(self.optimizer.learning_rate.numpy()), 0.1)
```
[the remark this answers is not in the document]
Comment (lior-dikstein) [reply tests/data_generation_tests/keras/test_scheduler_step.py:53]: In this test we just check the reduction
